x/ibc-hooks/keeper/keeper.go

```diff
@@ -32,31 +42,130 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(GetPacketAckKey(channel, packetSequence))
+}
+
+// DeriveIntermediateSender derives the sender address to be used when calling wasm hooks
 func DeriveIntermediateSender(channel, originalSender, bech32Prefix string) (string, error) {
 	senderStr := fmt.Sprintf("%s/%s", channel, originalSender)
 	senderHash32 := address.Hash(types.SenderPrefix, []byte(senderStr))
 	sender := sdk.AccAddress(senderHash32[:])
 	return sdk.Bech32ifyAddressBytes(bech32Prefix, sender)
 }
+
+// EmitIBCAck emits an event that the IBC packet has been acknowledged
+func (k Keeper) EmitIBCAck(ctx sdk.Context, sender, channel string, packetSequence uint64) ([]byte, error) {
+	contract := k.GetPacketAckActor(ctx, channel, packetSequence)
+	if contract == "" {
+		return nil, fmt.Errorf("no ack actor set for channel %s packet %d", channel, packetSequence)
+	}
+	// Only the contract itself can request for the ack to be emitted. This will generally happen as a callback
+	// when the result of other IBC actions has finished, but it could be exposed directly by the contract if the
+	// proper checks are made
+	if sender != contract {
+		return nil, fmt.Errorf("sender %s is not allowed to send an ack for channel %s packet %d", sender, channel, packetSequence)
+	}
+
+	// Write the acknowledgement
+	_, cap, err := k.channelKeeper.LookupModuleByChannel(ctx, "transfer", channel)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not retrieve module from port-id")
+	}
+
+	// Calling the contract. This could be made generic by using an interface if we want
+	// to support other types of AckActors, but keeping it here for now for simplicity.
+	contractAddr, err := sdk.AccAddressFromBech32(contract)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not parse contract address")
+	}
+
+	msg := types.IBCAsync{
+		RequestAck: types.RequestAck{RequestAckI: types.RequestAckI{
+			PacketSequence: packetSequence,
+			SourceChannel:  channel,
+		}},
+	}
+	msgBytes, err := json.Marshal(msg)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not marshal message")
+	}
+	bz, err := k.ContractKeeper.Sudo(ctx, contractAddr, msgBytes)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not execute contract")
+	}
+
+	ack, err := types.UnmarshalIBCAck(bz)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "could not unmarshal into IBCAckResponse or IBCAckError")
+
+	}
+	var newAck channeltypes.Acknowledgement
+	var packet exported.PacketI
+
+	switch ack.Type {
+	case "ack_response":
+		jsonAck, err := json.Marshal(ack.AckResponse)
+		if err != nil {
+			return nil, sdkerrors.Wrap(err, "could not marshal acknowledgement")
+		}
+		packet = ack.AckResponse.Packet
+		newAck = channeltypes.NewResultAcknowledgement(jsonAck)
+	case "ack_error":
+		packet = ack.AckError.Packet
+		newAck = osmoutils.NewEmitErrorAcknowledgement(ctx, types.ErrAckFromContract, ack.AckError.ContractError)
```

Comment:
Note to self: don't emit error acks but send result acks with information about being an error instead

This prevents the sender from releasing the escrowed tokens which could potentially cause a double spend